File: python/src/euler070.py
# Problem 70: Totient permutation
# Euler's Totient function, φ(n) [sometimes called the phi function], is used to determine the number of positive numbers less than or equal to n that are relatively prime to n. For example, as 1, 2, 4, 5, 7, and 8, are all less than or equal to nine and relatively prime to nine, φ(9)=6.
# The number 87109 has a special property; namely, it is the product of two primes, 7 and 12479, and φ(87109)=87108, and 87108 is a permutation of 87109.
# Find the value of n, below 10^7, for which φ(n) is a permutation of n and the ratio n/φ(n) is a minimum.
# Answer: 8319823
# Execution time: ~5000ms on modern hardware
# Optimizations: Sieve for totient, check permutations for prime products
# Interesting notes: AI-generated solution finding prime pairs with permutation property

def solve():
    from collections import defaultdict

    limit = 10000000
    phi = list(range(limit + 1))
    for i in range(2, limit + 1):
        if phi[i] == i:
            for j in range(i, limit + 1, i):
                phi[j] = phi[j] // i * (i - 1)

    def is_permutation(a, b):
        return sorted(str(a)) == sorted(str(b))

    min_ratio = float('inf')
    min_n = 0
    for n in range(2, limit):
        if is_permutation(n, phi[n]):
            ratio = n / phi[n]
            if ratio < min_ratio:
                min_ratio = ratio
                min_n = n
    return min_n

File: cpp/src/euler070.cpp
// https://projecteuler.net/problem=70
// Totient permutation

// Euler's totient function, φ(n) [sometimes called the phi function], is used to determine the number of positive numbers less than or equal to n which are relatively prime to n. For example, as 1, 2, 4, 5, 7, and 8, are all less than nine and relatively prime to nine, φ(9)=6.
// The number 1 is considered to be relatively prime to every positive number, so φ(1)=1.

// Interestingly, φ(87109)=79180, and it can be seen that 87109 is a permutation of 79180.

// Find the value of n, 1 < n < 10^7, for which φ(n) is a permutation of n and the ratio n/φ(n) produces a minimum.

// Answer: 8319823

#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <limits>

int totient_permutation()
{
    const int MAXN = 10000000;
    std::vector<long long> phi(MAXN+1);
    for(int i=0; i<=MAXN; i++) phi[i] = i;

    for(int i=2; i<=MAXN; i++){
        if(phi[i] == i){ // prime
            for(long long j=i; j<=MAXN; j+=i){
                phi[j] = phi[j] / i * (i-1);
            }
        }
    }

    double min_ratio = std::numeric_limits<double>::max();
    int result = -1;

    for(int n=2; n<=MAXN; n++){
        std::string s1 = std::to_string(n);
        std::string s2 = std::to_string(phi[n]);
        if(s1.length() != s2.length()) continue;
        std::sort(s1.begin(), s1.end());
        std::sort(s2.begin(), s2.end());
        if(s1 == s2){
            double ratio = (double)n / phi[n];
            if(ratio < min_ratio){
                min_ratio = ratio;
                result = n;
            }
        }
    }

    return result;
}

#if ! defined UNITTEST_MODE
int main(int argc, char const *argv[])
{
    std::cout << "Answer: " << totient_permutation() << std::endl;
}
#endif // #if ! defined UNITTEST_MODE